src/Lumi/Components/Form.purs

```diff
@@ -129,37 +138,175 @@ build editor = makeStateless (createComponent "Form") render where
              }
         contractProps = unsafeCoerce
 
-        fieldDivider = R.hr { className: "lumi field-divider" }
+     in defaultRenderForm
+          { readonly: props.readonly
+          , inlineTable: props.inlineTable
+          , forceTopLabels: props.forceTopLabels
+          }
+          forest
 
-        toRow = case _ of
-          Child { key, child } ->
-            maybe identity keyed key $ child
-          Wrapper { key, children } ->
-            R.div
-              { key: fromMaybe "" key
-              , children: [ intercalate fieldDivider (map toRow children) ]
-              }
-          Node { label, key, required, validationError, children } ->
-            maybe identity keyed key $ labeledField
-              { label: text body
-                  { children = [ label ]
-                  , className = toNullable (pure "field-label")
-                  }
-              , value: intercalate fieldDivider (map toRow children)
-              , validationError: validationError
-              , required: required
-              , forceTopLabel: forceTopLabels
-              , style: css {}
-              }
+-- | Render a form with state managed automatically.
+useForm
+  :: forall props unvalidated result xs
+   . RowToList unvalidated xs
+  => HMap ModifyValidated { | unvalidated } { | unvalidated }
+  => FormBuilder
+       { initialState :: { | unvalidated }
+       , readonly :: Boolean
+       , inlineTable :: Boolean
+       , forceTopLabels :: Boolean
```

Comment:
I don't see why we need `initialState`, `inlineTable` and `forceTopLabels` be props to the `FormBuilder`. These should be properties of the interpreter/renderer.

Reply:
I considered making them a separate argument. Would they ever be useful values for a form to pull out using `withProps`? Especially `initialState`. I don't think it really hurts though. The current `build` implementation does the same thing but hides it from the types, which is kind of dangerous.

Reply:
The main problem I see with this is that, besides being semantically confusing — the `FormBuilder` should be agnostic to what happens to it after its built/interpreted —, it restricts the types of forms that can be used with `useForm`. This would mean adding these props to all `FormBuilder`s we have, and it isn't relevant to the forms themselves but only to the rendering, which should be abstracted away inside `FormBuilder`s.

Having `initialState` as a prop assumes that the `FormBuilder` will be built with `{ | unvalidated }` as the initial state, but that's not true for its subforms.

The way `build` works is that it extends the record of props of the resulting _component_, not of the `FormBuilder`. These are two different sets of props: one contains the data needed for correctly specifying how the form should be rendered + the global properties of the form, and the other contains only the global properties of the form.

Reply:
Looking at the code it seems that `useForm` and `useForm'` would typecheck just fine taking a `FormBuilder { | props } { | unvalidated } result` as argument instead.

Reply:
That's true for `useForm'` but `useForm` depends on `defaultRenderForm`

Reply:
I really don't see why that's a problem, since `defaultRenderForm` doesn't take in a `FormBuilder` as argument.